subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -796,20 +789,35 @@ func generatePodScheduleResult(
 			PodPreemptInfo: &internal.PodPreemptInfo{VictimPods: victimPods},
 		}
 	} else {
-		// we check suggested nodes after the preemption is done, otherwise the preemption victims
+		// we find the selected node after the preemption is done, otherwise the preemption victims
 		// may cause the selected node to be excluded from the suggested nodes
-		// TODO: Keep selectedNode within suggestedNodeSet, so here should be Unreachable
-		if !suggestedNodeSet.Contains(selectedNode) && newGroup {
-			panic(fmt.Sprintf("[%v]: node %v picked by algorithm but not in K8S candidates",
-				internal.Key(pod), selectedNode))
+		affinityGroupBindInfo, selectedNode, selectedGpuIndices := generateAffinityGroupBindInfo(
+			groupPhysicalPlacement, groupVirtualPlacement, cellLevelToType, currentGpuNum, currentPodIndex, suggestedNodeSet)
+		if affinityGroupBindInfo == nil {
+			reason := "insufficient capacity in physical cluster"
+			if groupVirtualPlacement != nil {
+				reason = fmt.Sprintf("insufficient quota in VC %v", vc)
+			}
+			return internal.PodScheduleResult{
+				PodWaitInfo: &internal.PodWaitInfo{
+					Reason: reason,
+				},
+			}
+		}
+		if selectedNode == "" {
+			return internal.PodScheduleResult{
+				PodWaitInfo: &internal.PodWaitInfo{
+					Reason: fmt.Sprintf("node %v picked by algorithm but not in K8S candidates", selectedNode),
```

Comment:
>"node %v picked by algorithm but not in K8S candidates" [](start = 25, length = 55)

no node is picked here. Pls refine the msg :), such as:

insufficient capacity in physical cluster with respect to K8S candidates
/
insufficient quota in VC %v with respect to K8S candidates
?

Reply:
I used "insufficient capacity in physical cluster". "insufficient quota" seems to be not accurate because this case does not mean the user does not have quota :)

---
In reply to: [353695013](https://github.com/microsoft/pai/pull/3962#discussion_r353695013) [](ancestors = 353695013)

Reply:
I mean can we do similar thing like:

   reason := "insufficient capacity in physical cluster with respect to K8S candidates"
   if groupVirtualPlacement != nil {
    reason = fmt.Sprintf("insufficient quota in VC %v with respect to K8S candidates", vc)
   }

---
In reply to: [354076707](https://github.com/microsoft/pai/pull/3962#discussion_r354076707) [](ancestors = 354076707,353695013)